### ISSUE TITLE
FinTS dialog with Berliner Sparkasse aborts due to invalid dialog ID

### ISSUE THREAD

**Issue by cjolowicz** [raphaelm/python-fints#165]
**Describe the bug**
Minimal bootstrap fails with Sparkasse.

The FinTS dialog with Sparkasse aborts when fetching the tan medium (`client.get_tan_media`) with error `9010 - Die angegebene Bankreferenz/Dialog-ID ist nicht gültig.`

**Bank I tested this with*
Name of the bank: Berliner Sparkasse
FinTS URL: https://banking-be3.s-fints-pt-be.de/fints30

**Expected behavior**
Minimal bootstrap succeeds. 

I've successfully used this for a couple of years and I've passed a valid product ID. Last successful use was on April 27, 2024. First failure was on May 20, 2024.

**Code required to reproduce**
The code from `quickstart.rst` up to minimal bootstrap (`minimal_interactive_cli_bootstrap`).

**Log output / error message**
```
ERROR:fints.client:Dialog response: 9050 - Die Nachricht enthält Fehler.
ERROR:fints.client:Dialog response: 9800 - Dialog abgebrochen
ERROR:fints.client:Dialog response: 9010 - Die angegebene Bankreferenz/Dialog-ID ist nicht gültig.
```

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../raphaelm/python-fints/fints/utils.py", line 301, in minimal_interactive_cli_bootstrap
    m = client.get_tan_media()
        ^^^^^^^^^^^^^^^^^^^^^^
  File ".../raphaelm/python-fints/fints/client.py", line 1425, in get_tan_media
    response = method(dialog)(seg)
               ^^^^^^^^^^^^^^^^^^^
  File ".../raphaelm/python-fints/fints/dialog.py", line 156, in send
    self.client.process_response_message(self, response, internal_send=internal_send)
  File ".../raphaelm/python-fints/fints/client.py", line 230, in process_response_message
    self._process_response(dialog, None, response)
  File ".../raphaelm/python-fints/fints/client.py", line 1341, in _process_response
    raise FinTSClientError("Error during dialog initialization, could not fetch BPD. Please check that you "
fints.exceptions.FinTSClientError: Error during dialog initialization, could not fetch BPD. Please check that you passed the correct bank identifier to the HBCI URL of the correct bank.
```
**Additional context**
Python 3.11 macOS
python-fints 4.1.0

<summary>With earlier versions (4.0.0 and 3.1.0) I get this error instead (which I think was fixed in 4.1.0):

<details>

```
Traceback (most recent call last):
  File ".../raphaelm/python-fints/minimal.py", line 15, in <module>
    minimal_interactive_cli_bootstrap(f)
  File ".../raphaelm/python-fints/fints/utils.py", line 319, in minimal_interactive_cli_bootstrap
    if client.is_tan_media_required() and not client.selected_tan_medium:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../raphaelm/python-fints/fints/client.py", line 1180, in is_tan_media_required
    tan_mechanism = self.get_tan_mechanisms()[self.get_current_tan_mechanism()]
                    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: '999'
```
</details>
</summary>

My last successful use of this library was with a version before 4, likely 3.1.0.

**Comment by ntf1974** [comment]
Hello,

same behaviour on fetching Transactions from Sparkasse Pforzheim-Calw (BLZ: 66650085 / Endpoint: https://banking-bw3.s-fints-pt-bw.de/fints30).

Fetching failed immediately after changing TAN-Mechanism from QR to PushTAN (App).

Troubleshooting Script shows following error:
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

Traceback (most recent call last):
  File "/srv/its/bank/stmt.request/test/troubleshoot.py", line 21, in <module>
    minimal_interactive_cli_bootstrap(f)
  File "/srv/its/bank/stmt.request/venv/lib/python3.11/site-packages/fints/utils.py", line 301, in minimal_interactive_cli_bootstrap
    m = client.get_tan_media()
        ^^^^^^^^^^^^^^^^^^^^^^
  File "/srv/its/bank/stmt.request/venv/lib/python3.11/site-packages/fints/client.py", line 1425, in get_tan_media
    response = method(dialog)(seg)
               ^^^^^^^^^^^^^^^^^^^
  File "/srv/its/bank/stmt.request/venv/lib/python3.11/site-packages/fints/dialog.py", line 156, in send
    self.client.process_response_message(self, response, internal_send=internal_send)
  File "/srv/its/bank/stmt.request/venv/lib/python3.11/site-packages/fints/client.py", line 230, in process_response_message
    self._process_response(dialog, None, response)
  File "/srv/its/bank/stmt.request/venv/lib/python3.11/site-packages/fints/client.py", line 1341, in _process_response
    raise FinTSClientError("Error during dialog initialization, could not fetch BPD. Please check that you "
fints


Switching back TAN-Mechanism to QR will fix the error. Troubleshooting Script then works fine.

Also found similar notices in AqBanking Forum.

Chris

**Comment by cjolowicz** [comment]
There's also https://github.com/raphaelm/python-fints/issues/121 which may be related. It's the same error message, but with DKB.

**Comment by webffo** [comment]
Hello,
I'm having the same issue with Sparkasse Schwaben-Bodensee since today on BLZ 73150000 and URL https://banking-by5.s-fints-pt-by.de/fints30.

Any idea how this could be fixed?

**Comment by saddy001** [comment]
Same here since yesterday. I'm willing to test if someone is able to blind-fix this like we did in #155

**Comment by raphaelm** [comment]
#171 could also be related

**Comment by raphaelm** [comment]
Please try again with python-fints 4.2.0 and storing the system ID and passing it to the next run:
https://python-fints.readthedocs.io/en/latest/tans.html#system-ids

If the issue persists, feel free to reopen.

**Comment by ntf1974** [comment]
> Please try again with python-fints 4.2.0 and storing the system ID and passing it to the next run: https://python-fints.readthedocs.io/en/latest/tans.html#system-ids
> 
> If the issue persists, feel free to reopen.

I confirm: v4.2.0 fixes the problem for sparkasse pforzheim-calw (66650085). THANK YOU so much @raphaelm!

**Comment by saddy001** [comment]
Sadly doesn't work for me. If I pass the stored system_id to FinTS3PinTanClient it throws

> fints.exceptions.FinTSDialogInitError: Couldn't establish dialog with bank, Authentication data wrong?

What's unclear to me in the docs is "During your first connection, the bank will assign your client application a “system ID”". Because I get assigned a new system_id on every call to FinTS3PinTanClient(). I just stored the last one that I got assigned and used it for the next call. Am I doing it wrong?
